File: xophone/MIDIConnector.m
//
//  MIDIConnector.m
//  monoleap
//
//  Created by Dan Bar-Yaakov on 05/11/2015.
//  Copyright © 2015 Dan Bar-Yaakov. All rights reserved.
//

#import "MIDIConnector.h"
#import <CoreMIDI/CoreMIDI.h>

@implementation MIDIConnector {
    MIDIClientRef client;
    MIDIPortRef outputPort;
}

+ (id)sharedInstance {
    static MIDIConnector *sharedMyManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedMyManager = [[self alloc] init];
    });
    return sharedMyManager;
}

- (id)init {
    if (self = [super init]) {
        MIDIClientCreate(CFSTR("Monoleap"), nil, nil, &client);
        MIDIOutputPortCreate(client, CFSTR("MonoLeap Output Port"), &outputPort);
    }
    return self;
}

-(void)sendNoteOn:(int)noteNumber inChannel:(int)channel withVelocity:(int)velocity {
    Byte message[] = {0x90, noteNumber, velocity};
    [self sendMidi:message];
}

-(void)sendNoteOff:(int)noteNumber inChannel:(int)channel withVelocity:(int)velocity {
    Byte message[] = {0x80, noteNumber, 0};
    [self sendMidi:message];
}

-(void) sendControllerChange:(int)ccNumber value:(int)value inChannel:(int)channel {
    Byte message[] = {176, ccNumber, value};
    [self sendMidi:message];
}

-(void)sendMidi:(Byte*)message {
    MIDIPacketList packetList;
    MIDIPacket *packet = MIDIPacketListInit(&packetList);
    MIDIPacketListAdd(&packetList, sizeof(packetList), packet, 0, sizeof(message), message);
    ItemCount destinationCount = MIDIGetNumberOfDestinations();
    for (int i = 0; i < destinationCount; i++) {
        MIDISend(outputPort, MIDIGetDestination(i), &packetList);
    }
}



@end

File: xophone/MIDIConnector.swift
//
//  MIDIConnector.swift
//  monoleap
//
//  Created by Omer Elimelech on 10/09/2021.
//  Copyright © 2021 Dan Bar-Yaakov. All rights reserved.
//

import Foundation
import AudioKit

class MIDIConnector {
    
    static let instance = MIDIConnector()
    var midi = AudioKit.MIDI()
    
    init() {
        midi.openOutput()
    }
    
    func sendNote(on noteNumber: Int, inChannel channel: Int, withVelocity velocity: Int) {
        midi.openOutput()
        midi.sendEvent(MIDIEvent(noteOn: MIDINoteNumber(noteNumber), velocity: MIDIVelocity(127), channel: MIDIChannel(channel)))
    }
    
    public func sendNoteOff(_ noteNumber: Int, inChannel channel: Int, withVelocity velocity: Int) {
        midi.openOutput()
        midi.sendEvent(MIDIEvent(noteOff: MIDINoteNumber(noteNumber), velocity: MIDIVelocity(127), channel: MIDIChannel(channel)))
    }
    
    public func sendControllerChange(_ ccNumber: Int, value: Int, inChannel channel: Int) {
        midi.openOutput()
        midi.sendEvent(MIDIEvent(controllerChange: MIDIByte(ccNumber), value: MIDIByte(value), channel: MIDIChannel(channel)))
    }

}
